include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1134,6 +1134,72 @@ struct __early_exit_find_or
 // parallel_find_or - sync pattern
 //------------------------------------------------------------------------
 
+template <typename Tag>
+struct __parallel_find_or_nd_range_tuner
+{
+    // Tune the amount of work-groups and work-group size
+    template <typename _ExecutionPolicy>
+    std::tuple<std::size_t, std::size_t>
+    operator()(const _ExecutionPolicy& __exec, const std::size_t __rng_n) const
+    {
+        // TODO: find a way to generalize getting of reliable work-group size
+        // Limit the work-group size to prevent large sizes on CPUs. Empirically found value.
+        // This value exceeds the current practical limit for GPUs, but may need to be re-evaluated in the future.
+        const std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec, (std::size_t)4096);
+        std::size_t __n_groups = 1;
+        // If no more than 32 data elements per work item, a single work group will be used
+        if (__rng_n > __wgroup_size * 32)
+        {
+            // Compute the number of groups and limit by the number of compute units
+            __n_groups = std::min<std::size_t>(oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __wgroup_size),
+                                               oneapi::dpl::__internal::__max_compute_units(__exec));
+        }
+
+        return {__n_groups, __wgroup_size};
+    }
+};
+
+// No tuning for FPGA_EMU because we are not going to tune here the performance for FPGA emulation.
+#if !_ONEDPL_FPGA_EMU
+template <>
+struct __parallel_find_or_nd_range_tuner<oneapi::dpl::__internal::__device_backend_tag>
+{
+    // Tune the amount of work-groups and work-group size
+    template <typename _ExecutionPolicy>
+    std::tuple<std::size_t, std::size_t>
+    operator()(const _ExecutionPolicy& __exec, const std::size_t __rng_n) const
+    {
+        // Call common tuning function to get the work-group size
+        auto [__n_groups, __wgroup_size] = __parallel_find_or_nd_range_tuner<int>{}(__exec, __rng_n);
+
+        if (__n_groups > 1)
+        {
+            auto __iters_per_work_item =
+                oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __n_groups * __wgroup_size);
+
+            // If our work capacity is not enough to process all data in one iteration, will tune the number of work-groups
+            if (__iters_per_work_item > 1)
+            {
+                // Empirically found formula for GPU devices.
+                const auto __rng_x = __rng_n / 4096;
+                const auto __desired_iters_per_work_item = std::max(std::sqrt(__rng_x), 1.);
```

Comment:
We may want to add a TODO to revisit this question in the future when we are considering some more automated way of finding these empirical values / formulas.  

With the infrastructure we have now, I think this is a reasonable way of doing things, but I'm a little worried that we are just fitting to our benchmarking data and target hardware, and with a different set of data we might end up with a different formula here.  This is specifically interesting to these "early exit" routines because there can be such large variations in performance based on when the "early exit" is allowed by the data.

I don't think its any worse in this respect than before this PR so I don't object.  

With the addition of a TODO, I think I'm OK with the PR as it stands.